supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala

```diff
@@ -401,9 +431,41 @@ class TesAsyncBackendJobExecutionActor(override val standardParams: StandardAsyn
 
   override def requestsAbortAndDiesImmediately: Boolean = false
 
+  override def onTaskComplete(runStatus: TesRunStatus, handle: StandardAsyncPendingExecutionHandle): Unit = {
+    val taskEndTime = getTaskEndTime(handle, getTaskLogs)
+    if (runStatus == Error() | runStatus == Failed()) {
+      val errors = getErrorSeq(runStatus, handle, getErrorLogs)
+      errors.onComplete {
+        case Success(r) => tellMetadata(Map(CallMetadataKeys.Failures -> r))
+        case Failure(e) => log.error(e.getMessage)
+      }
+    }
+
+    taskEndTime.onComplete {
+      case Success(result) => tellMetadata(Map(CallMetadataKeys.TaskEndTime -> result))
+      case Failure(e) => log.error(e.getMessage)
+    }
+  }
+
   override def pollStatusAsync(handle: StandardAsyncPendingExecutionHandle): Future[TesRunStatus] =
+    handle.previousState match {
+      case None => pollStatus(handle, fetchCostData = true)
+      case Some(v) =>
+        val validatedCostData = for {
+          costData <- v.costData
+          startTime = costData.startTime
+          vmCost = costData.vmCost
+        } yield (startTime, vmCost)
+
+        validatedCostData match {
+          case Some(_) => pollStatus(handle, fetchCostData = false)
+          case None => pollStatus(handle, fetchCostData = true)
+        }
+    }
```

Comment:
I'm wondering if we can reorganize this a little to make it more concise and more obvious what the two options are. Is this equivalent to what you have?
```
handle.previousState match {
  case Some(TesRunStatus(_, _, Some(costData))) => pollStatus(handle, fetchCostData = false)
  case _ => pollStatus(handle, fetchCostData = true)
}
```